---
title: "R Notebook"
output: html_notebook
---

```{r}
library(tidyverse)
library(broom)
```

```{r}
df <- read_csv("gfp_regression_ecol_codon_use.csv")

df_ecol_0.7 <- read_csv("gfp_regression_speed.csv") %>%
  filter(speed == 2) %>%
  select(-speed) %>%
  mutate(ecol = 0.7)

df <- rbind(df, df_ecol_0.7)
```


```{r, fig.width=7, fig.height=4.5}
df %>% filter(charging_rate > 10) %>%
  filter(ecol %in% c(0.2, 0.4, 0.6, 0.8, 1)) %>%
  ggplot(aes(GFP, cellularProtein_norm, color = factor(codon))) +
    geom_point() + 
    geom_smooth(method = "lm", se = FALSE, formula = y ~ x + 1, fullrange=TRUE) +
    facet_grid(cols = vars(charging_rate), rows = vars(ecol)) +
    xlab("GFP expression") +
    ylab("norm e.coli expression") +
    scale_y_continuous(limits = c(0, 1)) +
    scale_x_continuous(breaks= c(5000 , 15000)) +
    #scale_x_continuous(limits=c(0, 10000)) + scale_y_continuous(limits=c(0.8, 1.0)) +
    scale_color_discrete_sequential(name="transgene \npref. codon\nuse (%)", palette = "viridis") +
    theme_bw()
```


```{r, fig.height=3, fig.width=9}
df %>% filter(charging_rate %in% c(100, 300, 1000, 3000, 10000), ecol == 0.8) %>%
  ggplot(aes(GFP, cellularProtein_norm, color = factor(codon))) +
    geom_point(size=1.5) + 
    geom_line() +
    #geom_line() +
    #geom_smooth(method = "lm", se = FALSE, formula = y ~ x + 1, fullrange=TRUE) +
    facet_wrap(~charging_rate, ncol=5) +
    xlab("transgene expression") +
    ylab("norm e.coli expression") +
    scale_y_continuous(limits = c(0.8, 1)) +
    scale_x_continuous(breaks= c(5000 , 15000)) +
    #scale_x_continuous(limits=c(0, 10000)) + scale_y_continuous(limits=c(0.8, 1.0)) +
    scale_color_discrete_sequential(name="transgene pref.\ncodon use (%)", palette = "viridis") +
    theme_bw() +
    theme(legend.position = "bottom")
```

```{r}
models <- df %>%
#    filter(codon != 0.5) %>%
    group_by(charging_rate, ecol) %>%
    do(tidy(lm(data = ., formula = cellularProtein_norm ~ factor(codon):GFP)))
models
```

```{r}
models %>%
  filter(term != "(Intercept)") %>% 
  mutate(term = recode(term, `factor(codon)0.1:GFP` = 0.1,
                       `factor(codon)0.3:GFP` = 0.3,
                       `factor(codon)0.5:GFP` = 0.5,
                       `factor(codon)0.7:GFP` = 0.7,
                       `factor(codon)0.9:GFP` = 0.9)) %>%
  group_by(ecol) %>%
  mutate(estimate = estimate*-1) %>%
  ungroup() %>%
  ggplot(aes(term, estimate, color=as.factor(charging_rate))) +
  facet_wrap(~ecol) +
  geom_point(size=3) +
  geom_smooth(method = "lm", se = FALSE) +
  scale_color_discrete_sequential(palette="SunsetDark", name="charging rate") +
  scale_x_continuous(name="transgene pref. codon use",
                     breaks=c(0.1, 0.3, 0.5, 0.7, 0.9)) +
  scale_y_continuous(name="change e.coli expression") +
  theme_bw()
```
```{r}
coef <- models %>%
  filter(term != "(Intercept)") %>% 
  mutate(term = recode(term, `factor(codon)0.1:GFP` = 0.1,
                       `factor(codon)0.3:GFP` = 0.3,
                       `factor(codon)0.5:GFP` = 0.5,
                       `factor(codon)0.7:GFP` = 0.7,
                       `factor(codon)0.9:GFP` = 0.9)) %>%
  group_by(charging_rate, ecol) %>%
  do(tidy(lm(data = ., formula = estimate ~ term))) %>%
  filter(term != "(Intercept)")

coef
```

```{r, fig.width=7, fig.height=4.5}
colors <- c( '#C4DDF8', '#4BABED', '#2A6B97', '#183044')

#colors <- c('#E0ECFA', '#9EBFDF', '#5986AA', '#1F4158')

#colors <- c('#CDE2F9', '#82B9E7', '#447FA9', '#1E4864')

midpoint <- 2.241834e-06
#midpoint <- 0

coef %>% filter(charging_rate > 10) %>%
  filter(ecol != 0.72) %>%
  #mutate(estimate = abs(estimate)) %>%
  ggplot(aes(x=charging_rate, y=estimate, color=as.factor(ecol))) + 
  #geom_hline(aes(yintercept = midpoint, linetype="midpoint"), linetype=2, show.legend = TRUE) +
  geom_point(size=2.5) +
  geom_line() +
  #xlim(1000, 20000) + ylim(0,1e-5) +
  scale_x_continuous(trans="log10", 
                     name="charging rate",
                     breaks=c(30, 100, 300, 1000, 3000, 10000),
                     labels=c("30", "100", "300", "1000", "3000", "10000"),
                     expand = expansion(mult = c(0.1, 0.1))) +
  scale_y_continuous(#trans="log10", 
                     name="recoding sensitivity", 
                     #limits = c(0.00000005, 0.0004),
                     expand = expansion(mult = c(0.05, 0.05))) +
  #scale_color_manual(values = colors, name = "e. coli codon use") +
  scale_color_discrete_qualitative(name="Fraction preferred\ncodons (cell)") +
  #coord_fixed(ratio=1) +
  theme_linedraw() +
  theme(axis.title = element_text(size = 12),
        axis.text = element_text(size = 12),
        legend.title = element_text(size=12),
        panel.background = element_rect(color="grey70"),
        panel.grid.minor.x = element_blank(),
        axis.line = element_line(color="black"),
        axis.ticks = element_blank())
```
```{r, fig.width=4.5, fig.height=4.5}
coef %>% 
  filter(charging_rate == 30) %>%
  filter(ecol %in% c(0.2, 0.4, 0.6, 0.7, 0.8, 1)) %>%
  ggplot(aes(x = ecol, y = estimate)) +
  geom_point() +
  geom_line() +
  geom_vline(xintercept = 0.73, linetype=2) +
  scale_x_continuous(name="Fraction preferred codons (cell)") +
  theme_bw()
```

```{r, fig.width=5.5, fig.height=4.5}
trna_count <- tibble(tRNA = c("ACG", "CCG", "UCU", "CCU", "ACG", "CCG", "UCU", "CCU"),
                     count = c(0.7, 0.1, 0.13, 0.07, 0.83, 0.08, 0.04, 0.08),
                     group = c("anticodon", "anticodon", "anticodon", "anticodon", "codon", "codon", "codon", "codon"))

trna_count %>%
  filter(group == "codon") %>%
  ggplot(aes(x = fct_reorder(tRNA, -count), y = count, fill=group)) +
  geom_col(position = "dodge") +
  theme_bw() +
  xlab("") +
  ylab("% total") +
  #ggtitle("threonine tRNA composition, E. coli")+
  scale_fill_manual(name = "", values = c("#01ad9a"), labels = c("cooresponding codon use,\nhighly expressed genes"))
  #theme(legend.position = "none")
```

